packages/gatsby/src/schema/types/type-defs.ts

```diff
@@ -0,0 +1,137 @@
+import { codeFrameColumns } from "@babel/code-frame"
+import report from "gatsby-cli/lib/reporter"
+import {
+  DocumentNode,
+  GraphQLError,
+  GraphQLEnumType,
+  GraphQLInterfaceType,
+  GraphQLList,
+  GraphQLNonNull,
+  GraphQLObjectType,
+  GraphQLOutputType,
+  GraphQLScalarType,
+  GraphQLUnionType,
+  Kind as GraphQLASTNodeKind,
+  parse,
+} from "graphql"
+import { GatsbyGraphQLType } from "../../schema/types/type-builders"
+import { dontInferExtensionName, inferExtensionName } from "../extensions"
+import { isGatsbyType } from "./type-builders"
+
+export type TypeOrTypeDef =
+  | string
+  | GraphQLOutputType
+  | GatsbyGraphQLType<any, any>
+
+export function isASTDocument(
+  typeOrTypeDef: TypeOrTypeDef | DocumentNode
+): typeOrTypeDef is DocumentNode {
+  return (
+    typeof typeOrTypeDef === `object` &&
+    !(typeOrTypeDef instanceof GraphQLScalarType) &&
+    !(typeOrTypeDef instanceof GraphQLObjectType) &&
+    !(typeOrTypeDef instanceof GraphQLInterfaceType) &&
+    !(typeOrTypeDef instanceof GraphQLUnionType) &&
+    !(typeOrTypeDef instanceof GraphQLEnumType) &&
+    !(typeOrTypeDef instanceof GraphQLList) &&
+    !(typeOrTypeDef instanceof GraphQLNonNull) &&
```

Comment:
This is required because [GraphQLScalarType](https://github.com/graphql/graphql-js/blob/v14.7.0/tstypes/type/definition.d.ts#L285), [GraphQLObjectType](https://github.com/graphql/graphql-js/blob/v14.7.0/tstypes/type/definition.d.ts#L371), [GraphQLInterfaceType](https://github.com/graphql/graphql-js/blob/v14.7.0/tstypes/type/definition.d.ts#L553), [GraphQLUnionType](https://github.com/graphql/graphql-js/blob/v14.7.0/tstypes/type/definition.d.ts#L619), [GraphQLEnumType](https://github.com/graphql/graphql-js/blob/v14.7.0/tstypes/type/definition.d.ts#L677), [GraphQLList](https://github.com/graphql/graphql-js/blob/v14.7.0/tstypes/type/definition.d.ts#L168) and [GraphQLNonNull](https://github.com/graphql/graphql-js/blob/v14.7.0/tstypes/type/definition.d.ts#L202) do not have `kind`